test/test_type_promotion.py

```diff
@@ -138,39 +139,53 @@ def test_mixed_type_backward(self):
-                    func(x, y).sum().backward()
-
-    # verifies that a.add(b) is the same as a.to(b.dtype).add(b) in cases
-    # where that should hold.
+        for func in [torch.add, torch.sub, torch.rsub, torch.mul, torch.div]:
+            for dtype1, dtype2 in itertools.product(f_dtypes, f_dtypes + i_dtypes):
+                x = torch.ones(10, requires_grad=True, dtype=dtype1, device=self.device)
+                y = torch.ones(10, dtype=dtype2, device=self.device)
+                func(x, y).sum().backward()
+
+    def _get_test_tensor(self, dtype, add_zeros=False, remove_zeros=False):
+        tensor = torch.randn([20, 20, 20], device=self.device) * 15
+        if add_zeros:
+            # add some zeros. This affects boolean tensors which would otherwise be all true
+            first[first < -15] = 0
+        tensor = tensor.to(dtype)
+        if remove_zeros:
+            # ensures no div-by-zero (with care for low precision uint8/half)
+            tensor[torch.abs(tensor) < 0.05] = 5
+        return tensor
+
+    # verifies that torch.<op>(first, second) is the same as 
+    # torch.<op>(first.to(common_dtype), second.to(common_dtype)) in cases where that should hold.
     def test_many_promotions(self):
-        from_to = {
-            torch.float16: torch.float32,
-            torch.half: torch.float16,
-            torch.int: torch.long,
-            torch.uint8: torch.long,
-            torch.uint8: torch.float,
-            torch.int: torch.float,
-            torch.int: torch.double,
-            torch.int16: torch.long,
-            torch.float16: torch.double,
-            torch.bool: torch.long,
-            torch.bool: torch.float
-        }
-
-        for k, v in from_to.items():
-            a = torch.rand([3, 3], device=self.device).to(k)  # no _th_uniform for half on cpu.
-            b = torch.rand([3, 3], device=self.device).to(v)
-            c = a.add(b)
-            d = a.to(v).add(b)
-            self.assertEqual(c.dtype, d.dtype, message='from {} to {}'.format(k, v))
-            self.assertEqual(c, d, message='from {} to {}'.format(k, v))
+        # Can also include half on CPU in cases where it will be promoted to a
+        # supported dtype
+        dtypes1 = torch.testing.get_all_math_dtypes('cuda')
+        dtypes2 = torch.testing.get_all_math_dtypes(self.device)
+        ops = [torch.add, torch.sub, torch.mul, torch.div, torch.rsub]
+        for dt1, dt2 in itertools.product(dtypes1, dtypes2):
+            for op, non_contiguous in itertools.product(ops, [True, False]):
+                common_dtype = torch.promote_types(dt1, dt2)
+                if common_dtype == torch.half and self.device == 'cpu':
+                    continue
+                if op == torch.sub and common_dtype != torch.bool:
+                    # Subtraction, the `-` operator, with a bool tensor is not supported.
+                    continue
+                first = self._get_test_tensor(dt1, dt1 == torch.bool)
```

Comment:
I think the add_zeros flag makes it harder to read the test code (what is the boolean argument doing?)

The solution add_zeros is trying to solve is that `torch.randn` is unable to produce a "random boolean tensor". In that case, we can abstract that away into the helper function, which is supposed to return us some valid test input. The function could like something like the following: (pseudocode)
```
def _get_test_tensor(self, dtype, remove_zeros=False):
    if dtype == torch.bool:
        return torch.randint(0, 1, shape, dtype=torch.bool, device=device)
    elif isintegral(dtype):
         return torch.randint(0, 15, shape, dtype=dtype, device=device)
    else:  # floating point
         return torch.randn(shape, dtype=dtype, device=device)
```
Let me know what you think

Reply:
it is context dependent though, not just based on the dtype. E.g. even if the dtype is bool, if the operator is div and it's the second argument we can't add zeros. I don't think an is_divisor flag would be more readable.

I can add the kwarg name at the call site if you think that would help, but to cover all options it's hard to move every condition into the function.

Reply:
I guess we're removing the zeros we add anyway in the div case so it still would work ignoring the op type for the add_zeros. If you think this makes a big readability difference I guess I can move it.

Reply:
I had three main concerns:
1. The readability of the boolean argument by itself.
2. Having an API with both "add_zeros" and "remove_zeros" is confusing -- which one wins out if both of them are True?
3. the redundancy between add_zeros and dtype because as you noted we are removing the zeros anyways in the div case.
4. the _get_test_tensor function does more work than it needs to. Instead of constructing a tensor with the desired values in one call, it constructs a tensor and then attempts to "hammer" it into the correct format.

On the other hand, all of this is just testing code and there are very few lines of it right now so we could just not make it better until it really becomes a problem

Reply:
> On the other hand, all of this is just testing code and there are very few lines of it right now so we could just not make it better until it really becomes a problem

I don't mind doing another round (working on something else right now). I think these have some value as-is though and we can also iterate. Some of the other tests in this file could use some cleanup too.